uber/models.py

```diff
@@ -494,93 +494,138 @@ def lookup_attendee(self, full_name, email, zip_code):
-                    self.shift_badges(badge_type, badge_num + offset, up=True)
+                if attendee.badge_num:
+                    offset = 1 if old_badge_type == attendee.badge_type and attendee.badge_num > (old_badge_num or 0) else 0
+                    self.shift_badges(attendee.badge_type, attendee.badge_num + offset, up=True)
+                attendee.badge_num = badge_num_keep
+
+            if not attendee.badge_num and self.needs_badge_num(attendee):
+                attendee.badge_num = self.get_next_badge_num(attendee.badge_type)
 
-            attendee.badge_num = badge_num
-            attendee.badge_type = badge_type
             return 'Badge updated'
 
+        def auto_badge_num(self, badge_type):
+            """
+            Gets the next available badge number for a badge type's range.
+
+            This ignores the session so that plugins can override the logic here if need be without worrying about
+            handling dirty sessions.
+
+            :param badge_type: Used as a starting point if no badges of the same type exist, and to select badges within
+            a specific range.
+            :return:
+            """
+            sametype = self.query(Attendee.badge_num).filter(Attendee.badge_type == badge_type,
+                                                   Attendee.badge_num >= c.BADGE_RANGES[badge_type][0],
+                                                   Attendee.badge_num <= c.BADGE_RANGES[badge_type][1])
+            if sametype.count():
+                sametype_list = [int(row[0]) for row in sametype.order_by(Attendee.badge_num).all()]
+
+                # Searches badge range for a gap in badge numbers; if none found, returns the latest badge number + 1
+                first = 0
+                last = sametype.count() - 1
+                if sametype_list[last] - sametype_list[first] == last:
+                    return sametype.order_by(Attendee.badge_num.desc()).first().badge_num + 1
+                else:
+                    middle = int((first+last)/2)
+
+                    while first < last:
+                        if (sametype_list[middle]-sametype_list[first]) != (middle - first):
+                            if (middle-first) == 1 and (sametype_list[middle]-sametype_list[first] > 1):
+                                return sametype_list[middle] - 1
+                            last = middle
+                        elif (sametype_list[last]-sametype_list[middle]) != (last-middle):
+                            if (last-middle) == 1 and (sametype_list[last]-sametype_list[middle] > 1):
+                                return sametype_list[middle] + 1
+                            first = middle
+                        else:
+                            return sametype.order_by(Attendee.badge_num.desc()).first().badge_num + 1
+
+                        middle = int((first+last)/2)
+            else:
+                return c.BADGE_RANGES[badge_type][0]
```

Comment:
I'm not super sure what logic you're referring to that guards against degenerate cases.

If you mean the gap-filling code, that's largely for events that don't want to shift custom badges (and that don't assign numbers at check-in, but rather let the system do so). What happens is that an admin decides, for whatever reason, to set an attendee's badge number to some ridiculously high number. If the system only looks at badge numbers higher than the highest number, we've suddenly 'lost' 500 badges or whatever. Anthrocon had to have their one-day badge range expanded in the middle of the event for exactly this reason.